lib/active_merchant/billing/gateways/quickbooks.rb

```diff
@@ -0,0 +1,312 @@
+        'PMT-1001' => STANDARD_ERROR_CODE[:invalid_cvc],   # payment processed, but cvc was invalid
+        'PMT-1002' => STANDARD_ERROR_CODE[:incorrect_address],   # payment processed, incorrect address info
+        'PMT-1003' => STANDARD_ERROR_CODE[:processing_error],   # payment processed, address info couldn't be validated
+
+        # Fraud Errors
+        'PMT-2000' => STANDARD_ERROR_CODE[:incorrect_cvc],      # Incorrect CVC
+        'PMT-2001' => STANDARD_ERROR_CODE[:invalid_cvc],        # CVC check unavaliable
+        'PMT-2002' => STANDARD_ERROR_CODE[:incorrect_address],  # Incorrect address
+        'PMT-2003' => STANDARD_ERROR_CODE[:incorrect_address],  # Address info unavailable
+
+        'PMT-3000' => STANDARD_ERROR_CODE[:processing_error],   # Merchant account could not be validated
+
+        # Invalid Request
+        'PMT-4000' => STANDARD_ERROR_CODE[:processing_error],   # Object is invalid
+        'PMT-4001' => STANDARD_ERROR_CODE[:processing_error],   # Object not found
+        'PMT-4002' => STANDARD_ERROR_CODE[:processing_error],   # Object is required
+
+        # Transaction Declined
+        'PMT-5000' => STANDARD_ERROR_CODE[:card_declined],      # Request was declined
+        'PMT-5001' => STANDARD_ERROR_CODE[:card_declined],      # Merchant does not support given payment method
+
+        # System Error
+        'PMT-6000' => STANDARD_ERROR_CODE[:processing_error],   # A temporary Issue prevented this request from being processed.
+      }
+
+      FRAUD_WARNING_CODES = ['PMT-1000','PMT-1001','PMT-1002','PMT-1003']
+
+      def initialize(options = {})
+        requires!(options, :consumer_key, :consumer_secret, :access_token, :token_secret, :realm)
+        @consumer_key = options[:consumer_key]
+        @consumer_secret = options[:consumer_secret]
+        @access_token = options[:access_token]
+        @token_secret = options[:token_secret]
+        @realm = options[:realm]
+        super
+      end
+
+      def purchase(money, payment, options = {})
+        post = {}
+        add_amount(post, money, options)
+        add_charge_data(post, payment, options)
+        post[:capture] = "true"
+
+        commit(ENDPOINT, post)
+      end
+
+      def authorize(money, payment, options = {})
+        post = {}
+        add_amount(post, money, options)
+        add_charge_data(post, payment, options)
+        post[:capture] = "false"
+
+        commit(ENDPOINT, post)
+      end
+
+      def capture(money, authorization, options = {})
+        capture_uri = "#{ENDPOINT}/#{CGI.escape(authorization)}/capture"
+
+        commit(capture_uri)
+      end
+
+      def refund(money, authorization, options = {})
+        post = {}
+        post[:amount] = money.is_a?(String) ? money : amount(money)
+        refund_uri = "#{ENDPOINT}/#{CGI.escape(authorization)}/refunds"
+
+        commit(refund_uri, post)
+      end
+
+      def void(authorization, options = {})
```

Comment:
I'm not really a fan of this `void` implementation - seems super complicated and error prone. I'd probably prefer to just not support `void` rather than trying to hack it as a refund.